### ISSUE TITLE
Legion: problem building on systems with installed complex.h

### ISSUE THREAD

**Issue by aheirich** [StanfordLegion/legion#773]
This problem came up in building TeraChem so I will use that as the example.  This could affect any Regent application when Regent is built and installed with cmake.

Here is a sample of the build error for main.cpp.  main.cpp includes eri-regent.h which includes legion.h.  

g++ -O2 -Wall -Werror -std=c++11 -I/usr/local/include -I/home/aheirich/Documents/SLAC/software/terachem/eri-regent/src/tests/cpp/../../../build/include -I/home/aheirich/Documents/SLAC/software/terachem/eri-regent/src/tests/cpp/../../cpp main.cpp /home/aheirich/Documents/SLAC/software/terachem/eri-regent/src/tests/cpp/../../cpp/eri_regent.cpp /home/aheirich/Documents/SLAC/software/terachem/eri-regent/src/tests/cpp/../../cpp/terachem_jdata.cpp /home/aheirich/Documents/SLAC/software/terachem/eri-regent/src/tests/cpp/../../cpp/terachem_kdata.cpp -o /home/aheirich/Documents/SLAC/software/terachem/eri-regent/src/tests/cpp/../../../build/eri_regent_test -L/usr/local/lib -lrealm -llegion -lregent -L/home/aheirich/Documents/SLAC/software/terachem/eri-regent/src/tests/cpp/../../../build/lib -lERIRegent -Wl,-rpath,/usr/local/lib,-rpath,/home/aheirich/Documents/SLAC/software/terachem/eri-regent/src/tests/cpp/../../../build/lib
In file included from /usr/local/include/legion.h:55,
                 from /home/aheirich/Documents/SLAC/software/terachem/eri-regent/src/tests/cpp/../../cpp/eri_regent.h:4,
                 from main.cpp:12:
/usr/local/include/legion/legion_redop.h:249:22: error: ‘complex’ was not declared in this scope
   class SumReduction<complex<float> > {
                      ^~~~~~~

This happens because Legion has its own version of complex.h which must be included.  On Ubuntu Linux this conflicts with /usr/include/c++/8/complex.h which gets pulled in by g++.  This causes the compilation error.

One way to resolve this would be to rename Legion's complex.h.  Why is there a file that clashes with a system header?

**Comment by aheirich** [comment]
a workaround is to disable complex reductions in regent by editing CMakeLists.txt

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 0e6f390e5..8a3e21bef 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -493,14 +493,14 @@ endif()
 # Complex support
 #------------------------------------------------------------------------------#
 
-if(Legion_BUILD_ALL OR Legion_BUILD_BINDINGS)
-  option(Legion_REDOP_COMPLEX "Use reduction operators for complex types" ON)
-else()
+#if(Legion_BUILD_ALL OR Legion_BUILD_BINDINGS)
+  #option(Legion_REDOP_COMPLEX "Use reduction operators for complex types" ON)
+#else()
   option(Legion_REDOP_COMPLEX "Use reduction operators for complex types" OFF)
-endif()
-if((Legion_BUILD_ALL OR Legion_BUILD_BINDINGS) AND NOT Legion_REDOP_COMPLEX)
-  message(FATAL_ERROR "Bindings require Legion_REDOP_COMPLEX to be set")
-endif()
+#endif()
+#if((Legion_BUILD_ALL OR Legion_BUILD_BINDINGS) AND NOT Legion_REDOP_COMPLEX)
+  #message(FATAL_ERROR "Bindings require Legion_REDOP_COMPLEX to be set")
+#endif()
 if(Legion_REDOP_COMPLEX)
   # define variable for legion_defines.h
   set(LEGION_REDOP_COMPLEX ON)

**Comment by streichler** [comment]
I think this will generally be a problem for building against an installed Legion, no matter whether it was built with cmake or the Makefile flow.  By calling the header "complex.h" and putting it in the root of the Legion include tree, an app will be forced to include it with
```#include <complex.h>```
and unfortunately, POSIX has already staked out that turf: https://pubs.opengroup.org/onlinepubs/9699919799/

@lightsighter if you had to choose between:
1) renaming the header to something like "legion_complex.h", or
2) moving the header into the legion include subdirectory, or
3) something else,
which would you choose?

**Comment by lightsighter** [comment]
1. I would be fine with renaming it to anything that doesn't have "legion" in the name of the file. It doesn't have anything to do with Legion. It's just an implementation of a templated complex type with support for CUDA. The only other header file that it depends on is the `half.h` header in the same directory. If we change the name of `complex.h` then we should also change the name of `half.h` accordingly.
2. I am against this for the same reason as above: the complex type doesn't have anything to do with Legion. It's also worth noting that at some point in the future, if Realm provides support for common reduction operators (+,*) then it will need to see this header file as well until it supports dynamic generation of reduction kernels for arbitrary types in the DMA system, so it should not go in the Legion directory.

**Comment by streichler** [comment]
Ah, I hadn't realized there was a second header file without a namespace...  Rather than sticking an essentially-arbitrary prefix on both filenames, I'd lean towards pushing them both into a subdirectory with a somewhat-meaningful-but-unlikely-to-conflict-with-other-packages name.  My suggestion would be "mathutils".  Thoughts?  Alternative suggestions?  (Vetoes are invalid without a replacement suggestion.)

**Comment by streichler** [comment]
I'd like to get a fix for this into the next release.  Any objections to moving the headers:
```
complex.h -> mathutils/complex.h
half.h    -> mathutils/half.h
```
?

**Comment by lightsighter** [comment]
It doesn't really feel like `mathutils`. More like `types` or `mathtypes`. There's no math functionality in there in the sense of algorithms, just types with standard operators.

**Comment by streichler** [comment]
Ok, let's go with `mathtypes` then.  `types` seems too general to be dropping into `/usr/local/` or the like without risking conflicts.

**Comment by lightsighter** [comment]
Although does this really solve the problem? The header is still called `complex.h`...

**Comment by lightsighter** [comment]
Or do you imagine the user having to write `#include mathtypes/complex.h`?

**Comment by aheirich** [comment]
I think you should change the name of the header

legion_complex.h


On Mar 20, 2020, at 2:53 PM, Mike Bauer <notifications@github.com<mailto:notifications@github.com>> wrote:



Or do you imagine the user having to write #include mathtypes/complex.h?

**Comment by lightsighter** [comment]
I already addressed that. The types don't have anything to do with Legion and they don't depend on Legion, so they shouldn't be prefixed with Legion. They are just generic implementations of a complex type and a half type with support for CUDA.

**Comment by streichler** [comment]
> `#include mathtypes/complex.h`?

Yes, the user will need to type
```
#include <mathtypes/complex.h>
```
because `#include <complex.h>` means "please include the POSIX-standard complex header".

**Comment by lightsighter** [comment]
Maybe we should call them `cudatypes` since their main feature is that they support CUDA. Although the half one is really weird because `half_t` is already a type in CUDA, we're just providing equivalent support for the same type on the CPU.

**Comment by streichler** [comment]
This is in master with commit 42b4dd21.  I went with `mathtypes` in anticipation of future support for other (non-CUDA) accelerators as well.
@aheirich can you undo your workaround and see if you can build with these changes?

**Comment by lightsighter** [comment]
@aheirich Can you confirm that this issue is fixed? We would like to get the issue closed for the 20.03 release.

**Comment by aheirich** [comment]
solved in master
